agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -1258,49 +1251,81 @@ bool backhaul_manager::send_autoconfig_search_message(std::shared_ptr<sRadioInfo
+    auto mac_comp = [](const sMacAddr &addr1, const sMacAddr &addr2) -> bool {
+        return memcmp((char *)addr1.oct, (char *)addr2.oct, sizeof(sMacAddr)) < 0;
+    };
+
+    // sort the given bssid list
+    std::vector<sMacAddr> given_bssid_list = bssid_list;
+    std::sort(given_bssid_list.begin(), given_bssid_list.end(), mac_comp);
+
+    // for each socket - send only relevant bssids
+    auto ret = true;
     for (auto socket : slaves_sockets) {
-        for (const auto &mac : bssid_list) {
-            int i = 0;
-            if (mac == socket->vaps_list.vaps[i].mac) {
-                LOG(DEBUG) << "Forwarding AP_METRICS_QUERY_MESSAGE message to son_slave, bssid: "
-                           << std::hex << tlvf::mac_to_string(mac);
-
-                auto forward =
-                    cmdu_tx.create(mid, ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE);
-                if (!forward) {
-                    LOG(ERROR) << "Failed to create AP_METRICS_QUERY_MESSAGE";
-                    return false;
-                }
 
-                auto query = cmdu_tx.addClass<wfa_map::tlvApMetricQuery>();
-                if (!query) {
-                    LOG(ERROR) << "Failed addClass<wfa_map::tlvApMetricQuery>";
-                    return false;
-                }
+        // create the message to forward to current slaves
+        auto forward = cmdu_tx.create(mid, ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE);
+        if (!forward) {
+            LOG(ERROR) << "Failed to create AP_METRICS_QUERY_MESSAGE";
+            return false;
+        }
 
-                if (!query->alloc_bssid_list(1)) {
-                    LOG(ERROR) << "Failed allocate memory for bssid_list";
-                    return false;
-                }
+        // copy current vaps bssid
+        std::vector<sMacAddr> current_socket_bssids;
+        std::transform(socket->vaps_list.vaps, socket->vaps_list.vaps + beerocks::IFACE_TOTAL_VAPS,
+                       std::back_inserter(current_socket_bssids),
+                       [](const beerocks_message::sVapInfo &vap_info) { return vap_info.mac; });
 
-                auto list         = query->bssid_list(0);
-                std::get<0>(list) = true;
-                std::get<1>(list) = mac;
+        // sort current vaps bssid
+        std::sort(current_socket_bssids.begin(), current_socket_bssids.end(), mac_comp);
 
-                if (!message_com::send_cmdu(socket->slave, cmdu_tx)) {
-                    LOG(ERROR) << "Failed forwarding AP_METRICS_QUERY_MESSAGE message to son_slave";
-                    ret = false;
-                    continue;
-                } else {
-                    ret = true;
-                    // Fill a query vector
-                    m_ap_metric_query.push_back({socket->slave, mac});
-                }
-            }
+        // fill a list of the common bssids: given list with current vaps
+        std::vector<sMacAddr> required_bssids;
```

Comment:
Begging to differ here...

std::vector is most efficient for small containers, especially if you only append at the end, not in the middle. Reallocation doesn't actually happen that often, because it will over-allocate a bit. A std::list, for instance, will usually have a lot more allocations. A deque will have slightly less allocations, but use a lot more space. Also, even though both vector and deque have O(1) access, the deque access takes twice as much time (it needs two pointer dereferences instead of just one).

So, the only good reasons to use a deque are:
- you did profiling and it turns out to be faster for realistic use cases;
- the semantics we need are more like a deque than a general container.

I think neither is true here.

Reply:
I kind of getting the feeling that `std::vector` is kind of the default container everyone uses without giving much thought if it the right container for the job.
It is good for use if you push or insert elements once or very few times. This is not the case here.

Here is a link to a flow chart that explains how to choose a container.
I think we should adopt it, and add it to our coding style document.

Reply:
I agree that we should adopt something.

However, I'm kind of a fan of the flow chart from [CppGuideline SL.con.2](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#slcon2-prefer-using-stl-vector-by-default-unless-you-have-a-reason-to-use-a-different-container): Use std::array if size is constant, std::vector if not. The implicit corollary is: only use one of the other containers if profiling proves it's more efficient. I like that as a programmer, I don't have to think about it but I can just use std::vector. Simpler is better IMO. But I'm not against using the flow chart either.

Note that for this particular case, the flow chart will point to std::vector (order is not important, no key lookup, no persistent positions, size doesn't vary wildly) (although I'm not entirely sure what "size varies wildly" means).

Reply:
"Size doesn't vary wildly" means that you push or insert elements once or a few times.
In this case, the vector will be filled each time handling the message, so its size varies wildly. 
Also, a case when you push elements one by one is a case of "size varies wildly" (Unless you know how much you are going to push and reserve the size first).

I don't agree with what is said on "CppGuideline SL.con.2". It is too general and could miss bigtime.